code/site/components/com_pages/http/client.php

```diff
@@ -0,0 +1,347 @@
+ * @license     GNU GPLv3 <http://www.gnu.org/licenses/gpl.html>
+ * @link        https://github.com/joomlatools/joomlatools-pages for the canonical source repository
+ */
+
+/**
+ * Http Client Cache
+ *
+ * This class implement a http cache following the https://tools.ietf.org/html/rfc7234 specification.
+ *
+ * Features:
+ *   - Cache Validation using ETag and Last-Modified headers
+ *   - Cache Invalidation of none-safe requests
+ *   - Freshening Stored Responses upon Validation
+ *   - Freshening Responses via HEAD
+ *   - Serving Stale Responses
+ *
+ * Limitations:
+ *   - Expires header (deprecated in HTTP/1.1)
+ *   - Vary header
+ *   - Cache-Control directives
+ *       - request: max-stale
+ *
+ * The cache isn't able to send stale responses. If the cache encounters an error trying to validate or refresh
+ * itself the cache will throw an KHttpException
+ *
+ * The cache has a `cache_force`setting that when enabled will disregard response Cache-Control directives and
+ * ETag and Last-Mofified headers. Instead the cache will fallback to the `cache_time` to establish freshness
+ *
+ * @author  Johan Janssens <https://github.com/johanjanssens>
+ */
+class ComPagesHttpClient extends KHttpClient
+{
+    //The resource was served from cache
+    const CACHE_HIT     = 'HIT';
+
+    //The resource was not found in cache.
+    const CACHE_MISS    = 'MISS';
+
+    //The origin server doesn't allow to cache the resource.
+    const CACHE_BYPASS  = 'BYPASS';
+
+    //The resource was found in cache but has since expired. It has been fetched from the origin server
+    const CACHE_EXPIRED = 'EXPIRED';
+
+    //The resource was found in cache and couldn't be validated (through If-Modified-Since or If-None-Match header). It
+    //has been fetched from the origin server
+    const CACHE_MODIFIED  = 'MODIFIED';
+
+    //The resource was found in cache and was validated (through If-Modified-Since or If-None-Match header). It has been
+    //served from the cache
+    const CACHE_REFRESHED  = 'REFRESHED';
+
+    //The origin server couldn't be reached to validate the cache, a stale resource was served from cache instead
+    const CACHE_STALE  = 'STALE';
+
+    protected function _initialize(KObjectConfig $config)
+    {
+        $config->append([
+            'cache'       => false,
+            'cache_path'  => $this->getObject('com://site/pages.config')->getSitePath('cache'),
+            'cache_time'  => 60*60*24, //1 day https://tools.ietf.org/html/rfc7234#section-4.2.2
+            'cache_force' => false,
+            'debug'       => JDEBUG ? true : false,
+        ]);
+
+        parent::_initialize($config);
+    }
+
+    public function send(KHttpRequestInterface $request)
+    {
+        $url = $request->getUrl();
+
+        if($request->isCacheable())
+        {
+            //Cache exists
+            if($file = $this->isCached($url, false))
+            {
+                if (!$cache = require($file)) {
```

Comment:
@johanjanssens should we not invalidate the cache here? Otherwise there is no coming back from the error no?

Reply:
@ercanozkaya Good catch, fixed.